tests/unittests/tests-core/tests-core-clist.c

```diff
@@ -70,36 +173,107 @@ static void test_clist_remove_two(void)
 
     test_clist_add_two();
 
-    clist_remove_head(list);
-    clist_remove_head(list);
+    clist_lpop(list);
+    clist_lpop(list);
 
     TEST_ASSERT_NULL(list->next);
 }
 
-static void test_clist_advance(void)
+static void test_clist_lpoprpush(void)
 {
     list_node_t *list = &test_clist;
     list->next = NULL;
 
     test_clist_add_two();
 
-    clist_advance(list);
+    clist_lpoprpush(list);
 
     TEST_ASSERT(list->next->next == &tests_clist_buf[1]);
 
-    clist_advance(list);
+    clist_lpoprpush(list);
 
     TEST_ASSERT(list->next->next == &tests_clist_buf[0]);
 }
 
+static int _foreach_called;
+static int _foreach_visited[TEST_CLIST_LEN];
+static int _foreach_abort_after = TEST_CLIST_LEN/2;
+
+static void _foreach_test(clist_node_t *node)
+{
+    TEST_ASSERT(node == &tests_clist_buf[_foreach_called]);
+
+    for (int i = 0; i < TEST_CLIST_LEN; i++) {
+        if (node == &tests_clist_buf[i]) {
+            _foreach_visited[i]++;
+            break;
+        }
+    }
+
+    for (int i = 0; i < TEST_CLIST_LEN; i++) {
+        if (i <= _foreach_called) {
+            TEST_ASSERT(_foreach_visited[i] == 1);
+        }
+        else {
+            TEST_ASSERT(_foreach_visited[i] == 0);
+        }
+    }
+
+    _foreach_called++;
+}
+
+/* embunit test macros only work within void returning functions, so this
+ * trampoline function is needed */
+static int _foreach_test_trampoline(clist_node_t *node)
+{
+    _foreach_test(node);
+    if (_foreach_called == _foreach_abort_after) {
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+static void test_clist_foreach(void)
+{
+    set_up();
```

Comment:
clist needs to be clear before the test starts, but this test doesn't call one of the _add_n() functions.